### PR TITLE
Switch to existing tab with the same workspace

### PR DESCRIPTION
**Pull request by akurinnoy**
### What does this PR do?

When opening a workspace, the dashboard looks for an already open tab with the same workspace. If a tab exists, the dashboard focuses it without creating a new tab.

### What issues does this PR fix or reference?

resolves https://github.com/eclipse-che/che/issues/22933

### Is it tested? How?

1. Create a new workspace, or open an existing one. A new browser tab should be created.
2. Go back to the dashboard tab, and
    - click the same workspace in the Recent Workspaces section, or
    - click "Open" on the same workspace in the workspaces list, or
    - open the workspace actions menu for the same workspace, and click the "Open" item
3. The existing workspace tab should be focused.

### CODE REVIEW

**Review by ibuziuk — APPROVED** [review]

File: packages/dashboard-frontend/src/services/helpers/location.ts [hunk 1]
```diff
@@ -105,5 +105,6 @@ function _buildLocationObject(pathAndQuery: string): Location {
 }
 
 export function toHref(history: History, location: Location): string {
-  return history.createHref(location);
+  const fragment = history.createHref(location);
+  return window.origin + window.location.pathname + fragment;
```
[the remark this answers is not in the document]
Comment (olexii4) [reply packages/dashboard-frontend/src/services/helpers/location.ts:109]: In some cases `window.origin` could return the string "null".

@akurinnoy WDYT?

**Review by olexii4 — APPROVED** [review]
I have tested these changes with Minikube. Everything looks like it was expected.

File: packages/dashboard-frontend/src/services/helpers/location.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (akurinnoy) [reply packages/dashboard-frontend/src/services/helpers/location.ts]: good catch, thanks!

**Comment by dmytro-ndp** [comment]
@akurinnoy : I have followed test scenario from description and failed on step 2:
> 2. Go back to the dashboard tab, and
> - click the same workspace in the Recent Workspaces section, or
> - click "Open" on the same workspace in the workspaces list, or
> - open the workspace actions menu for the same workspace, and click the "Open" item

There was no new workspace in the **Workspaces** list, of **RECENT WORKSPACES** section of left panel after the new empty workspace has started:
Screenshot from 2024-05-15 19-11-31

**Comment by akurinnoy** [comment]
@dmytro-ndp can you please test this PR one more time?

**Comment by dmytro-ndp** [comment]
@akurinnoy: thank you for the fixing relogin case.
Now Dashboard switches user to existing workspace tab, but redirecting to start workspace page instead of continue working in the same Editor window after the re-login. Not sure if it's expected behavior in case the user has uncommitted changes in the workspace, and so could lose them.
Screencast: [relogin-2024.05.17-19_47_33.webm](https://github.com/eclipse-che/che-dashboard/assets/1197777/24118ffd-bd9b-4496-b120-7fe8618cdcbb)

I have also identified 2 more test cases:
#### Case "Open starting workspace again"
1. Start new workspace creation in new tab.
2. Go back to the dashboard tab before workspace is started and Editor is opened, and
- click the same workspace in the Recent Workspaces section, or
- click "Open" on the same workspace in the workspaces list, or
- open the workspace actions menu for the same workspace, and click the "Open" item
3. ~~The existing workspace tab should be focused.~~ Dashboard created new tab
Screencast: 
[create-several-workspaces-at-a-time-2024.05.17-19_45_07.webm](https://github.com/eclipse-che/che-dashboard/assets/1197777/83f1bc55-cce9-4950-bd10-8af3cc552931)

#### Case "Workspace with IntellijIDEA Community Editor" 
1. Create a new workspace with IntellijIDEA Community Editor.
2. Go back to the dashboard tab before workspace is started and Editor is opened, and
- click the same workspace in the Recent Workspaces section, or
- click "Open" on the same workspace in the workspaces list, or
- open the workspace actions menu for the same workspace, and click the "Open" item
3. ~~The existing workspace tab should be focused.~~ Dashboard was creating new tab first few seconds.
Screencast: 
[intellijIDEA-2024.05.17-20_07_03.webm](https://github.com/eclipse-che/che-dashboard/assets/1197777/7afe4803-82bc-453e-8166-54b6165ee1de)

I have also noticed https://issues.redhat.com/browse/CRW-6287 error:
Screenshot from 2024-05-17 19-37-25

**Comment by akurinnoy** [comment]
Hi @dmytro-ndp ,

> Now Dashboard switches user to existing workspace tab, but redirecting to start workspace page instead of continue working in the same Editor window after the re-login. Not sure if it's expected behavior in case the user has uncommitted changes in the workspace, and so could lose them.

This has been the case for quite some time, and this PR doesn't change that behavior. 
During the re-login, the Dashboard page loses references to the previously opened tabs/windows and cannot set focus on them. In this case, the dashboard can only reuse the existing workspace tab (found by the window context string) to open the necessary workspace in it.

> Case "Open starting workspace again"

> Case "Workspace with IntellijIDEA Community Editor"

The latest fixup commit should fix these cases.

**Comment by dmytro-ndp** [comment]
@akurinnoy : thanks for the addressing review notices!

> > Case "Open starting workspace again"
> 
> > Case "Workspace with IntellijIDEA Community Editor"
> 
> The latest fixup commit should fix these cases.

I can confirm that those cases are working as expected in the fresh User Dashboard deployed from this PR (quay.io/eclipse/che-dashboard@sha256:b6e591b610156484f1eb6195293c575c1d6f2487198efcadad3f93aa7598a229).
Screencast: [screencast-bpconcjcammlapcogcnnelfmaeghhagj-2024.05.22-18_37_12.webm](https://github.com/eclipse-che/che-dashboard/assets/1197777/480996b2-0b7f-426d-9ede-8a36c1a0a0a4)

I have only noticed that it was needed to click several times on workspace name in the left left panel of Dashboard to open redirect to the new tab.

**Review by dmytro-ndp — APPROVED** [review]

**Pull request merged**